docs/cardano-monads.rst

```diff
@@ -0,0 +1,608 @@
+of our custom classes yet. Just to clarify: in this section we'll discuss the
+current transitional state, and it's more painful than what was actually
+proposed by FPComplete.
+
+In other words, there is a compliсation that actually we have several approaches
+combined:
+
+* General approach a la ExecMode
+* Mockable-based constraints
+* Reflection-based constants (used instead of MonadReader for constant data within single execution)
+* Method dictionaries (for supporting SendActions, a primitive from networking)
+* ReaderT with method dictionaries
+
+General Approach
+~~~~~~~~~~~~~~~~
+
+Now our code follows this pattern::
+
+    -- effect definitions
+
+    class MonadBaz
+        baz :: ...
+
+    defaultBaz = ...
+
+    class MonadZaz
+        zaz :: ...
+
+    defaultZaz = ...
+
+
+    -- mode definitions
+
+    data QuuxCtx = ...
+
+    type QuuxMode = ReaderT QuuxCtx Production
+
+    instance HasFoo ModeCtx
+    instance HasBar ModeCtx
+
+    instance MonadBaz QuuxMode
+        baz = defaultBaz
+
+    instance MonadZaz QuuxMode
+        zaz = defaultZaz
+
+
+Pros:
+
+* Good compile-time performance. There's only a single layer of transformers,
+  instance search is quick, no coercions involved.
+
+Cons:
+
+* Bad run-time configurability. (Same as above)
+
+* Boilerplate. Various ``HasFoo`` instances with field lenses, ``MonadBaz``
+  instances to choose method implementations for the current mode. There's also
+  that annoying ``ctx`` parameter.
+
+* Cost of introducing new modes. The approach is inflexible, as introducing a
+  new mode has an extremely high cost (due to boilerplate). Assuming we want to
+  avoid nested ``ReaderT``, adding one more field to the context requires a new
+  mode.
+
+* Lack of inheritance. It's hard to define one mode in terms of another, with
+  only minor changes. Either it becomes hard to maintain consistency, or it
+  becomes hard to do overrides (as happened to ``AuxxMode``).
+
+* Volatile ``runReaderT``. It's difficult to reason about code when
+  ``runReaderT`` might imply something besides supplyng the value of the
+  ``ReaderT`` environment (handling ``MonadReader``). The situation arises
+  because we define instances for ``ReaderT`` without a newtype (and sometimes
+  even with ``{-# OVERLAPPING #-}``).
+
+Conclusion:
+
+* Current solution requires huge swaths of boilerplate code, it's hard to
+  reason about the code, and it's inflexible. We must seek other options.
+
+Usage of ``reflection``
+~~~~~~~~~~~~~~~~~~~~~~~
+
+To pass constant configuration to application components, instead of
+``ReaderT`` we use ``reflection``. More specifically, we use the dumb and unsafe
+``Given``-style reflection, avoiding the type-level complications of proper
+``Reifies``-style reflection. It turned out to be a great design choice: we've
+cut the potential amount of custom classes greatly, and the configs are
+available even in class instances (such as ``Bi``).
```

Comment:
I should mention that using the configuration options in type class instances really begs this solution. If we didn't need to do that, then we could just pass the configuration in a reader context, which is of course one of the flagship applications of a reader transformer.

Reply:
Another major advantage of `Given` is that it can be used in non-monadic code. The only reason to use `ReaderT` is if one needs `local`.

Reply:
Does `Given` approach have the same performance as `ReaderT` approach? Does it work well if there are different multiple `give`s (for different types, as we have now)?

Reply:
> Does Given approach have the same performance as ReaderT approach?

I believe the performance of `Given` should be equal or better than that of `ReaderT`, since the compiler makes more assumptions about coherency and applies more aggressive optimizations. Source: this Kmett's reddit comment https://www.reddit.com/r/haskell/comments/3hw90k/what_is_the_reflection_package_for/cub65fp/

> Does it work well if there are different multiple gives (for different types, as we have now)?

Yes, I believe the runtime representations of `(Given a, Given b)` and `Given (a, b)` are equivalent.

I don't have benchmarks to back up this claim, though.